### PR TITLE
Add parameterized docker flag support for custom build android package script.

### PR DESCRIPTION
**Pull request by yeziyqf1**
### Description

This change adds parameterized docker run flag support for custom build android package script. 
p.s. This Android custom build script isn't included in the release. it's intended for usage by a user wanting to do a custom build. This script is not included in the python package of this release.

### Motivation and Context

Currently the suggested approach building custom Android ONNXRuntime library is based on the following script:
https://github.com/microsoft/onnxruntime/blob/main/tools/android_custom_build/build_custom_android_package.py

In this script, the docker run settings are set to require an interactive environment, with the flag "-it". Thus if running in a non-interactive environment, this would be an issue.

This change adds parameterized docker run flag support, so user can specify whether to run docker in an interactive/non-interactive environment. If the user doesn't specify the content of "--docker_interactive_mode_flag", the default behavior is the same as what has been set previously: Use flag "-it" to make it run in an interactive environment. If user wants to run this script non-interactively, they can specify with the command "--docker_interactive_mode_flag i" to make it run non-interactively.

The parameter is added along with description on how it would work, and suggested use cases.

Update: In the discussion under this PR, we decided to remove the `-it` flag in the end, as it is not necessary for working in a console (interactive envinironment) and works with building in non-interactive environment as well.

### CODE REVIEW

**Comment by snnn** [comment]
But, is the flag "-it" ever needed?

**Comment by yeziyqf1** [comment]
> 

@snnn For your question, maybe the section author would have a better understanding on weather this flag is needed. @edgchen1 Regards the flag docker run flag "-it". could you give us some insights on how it is needed, and weather this change looks appropriate to you? Thanks!

**Comment by yeziyqf1** [comment]
> But, is the flag "-it" ever needed?

From docker documentation, it seems like for interactive processes (like a shell), we must use "-i" "-t" together in order to allocate a tty for the container process. 

Thus, if a normal user wants to build it from command line, most likely it would be required to use "-it" flag.
@edgchen1 Please correct me if I'm wrong. Thx

image

**Comment by edgchen1** [comment]
> > But, is the flag "-it" ever needed?
> 
> From docker documentation, it seems like for interactive processes (like a shell), we must use "-i" "-t" together in order to allocate a tty for the container process. https://docs.docker.com/engine/reference/run/
> 
> Thus, if a normal user wants to build it from command line, most likely it would be required to use "-it" flag. @edgchen1 Please correct me if I'm wrong. Thx
> 
> image

Sorry, I am not sure whether it's actually required. Perhaps not. Would suggest trying it out in both an interactive and non-interactive environment.

From here: https://docs.docker.com/engine/reference/run/#foreground
"It can even pretend to be a TTY (this is what most command line executables expect) and pass along signals."

It would be good if signals still work correctly in the interactive case, e.g., so one can stop it with Ctrl+C.

File: tools/android_custom_build/build_custom_android_package.py
Comment (edgchen1) [inline tools/android_custom_build/build_custom_android_package.py:148]: this offers a bit too much flexibility - a user can pass an arbitrary value and perhaps specify other docker options.
if you want to make this configurable with an option, would suggest just having that option control whether "-it" or "-i" is used.

Comment (skottmckay) [reply tools/android_custom_build/build_custom_android_package.py:148]: I agree. If the choice is between `-it` and `-i` it's about whether the psuedo TTY is created isn't it, so maybe an arg called docker_psuedo_tty with default of True to match current behaviour?

Comment (yeziyqf1) [reply tools/android_custom_build/build_custom_android_package.py:148]: This proposal sounds good, I will make this argument as a bool so we only change between the "-it" and "-i" option.

**Comment by yeziyqf1** [comment]
> > > But, is the flag "-it" ever needed?
> > 
> > 
> > From docker documentation, it seems like for interactive processes (like a shell), we must use "-i" "-t" together in order to allocate a tty for the container process. https://docs.docker.com/engine/reference/run/
> > Thus, if a normal user wants to build it from command line, most likely it would be required to use "-it" flag. @edgchen1 Please correct me if I'm wrong. Thx
> > image
> 
> Sorry, I am not sure whether it's actually required. Perhaps not. Would suggest trying it out in both an interactive and non-interactive environment.
> 
> From here: https://docs.docker.com/engine/reference/run/#foreground "It can even pretend to be a TTY (this is what most command line executables expect) and pass along signals."
> 
> It would be good if signals still work correctly in the interactive case, e.g., so one can stop it with Ctrl+C.

I will remove this "-it" option and see how that works in both interactive and interactive cases, and let you know.

**Comment by yeziyqf1** [comment]
> > 
> 
> I've gave the removed "-it" option a try, the local build in an interactive environment works fine, and we can stop the build anytime using CTRL-C. However, in a non-interactive environment, the build would start looping when reached `Step 10/18 : RUN ${ANDROID_HOME}/cmdline-tools/latest/bin/sdkmanager --install "platforms;android-32" "ndk;${NDK_VERSION}"`, as it is waiting for the user response for weather accepting the licenses. As it receives no response, it starts over again and again.
> 
> So, I would propose we go with the proposal we discussed above: To make an arg called `docker_psuedo_tty` with default of True to match current behavior, so we only change between the "-it" and "-i" option.
> 
> @snnn @edgchen1 @skottmckay FYI.

I've pushed a new commit addressing the solution we discussed above. Please review when convenient, thanks all!

**Comment by edgchen1** [comment]
> > 
> 
> I've gave the removed "-it" option a try, the local build in an interactive environment works fine, and we can stop the build anytime using CTRL-C. However, in a non-interactive environment, the build would start looping when reached `Step 10/18 : RUN ${ANDROID_HOME}/cmdline-tools/latest/bin/sdkmanager --install "platforms;android-32" "ndk;${NDK_VERSION}"`, as it is waiting for the user response for weather accepting the licenses. As it receives no response, it starts over again and again.
> 
> So, I would propose we go with the proposal we discussed above: To make an arg called `docker_psuedo_tty` with default of True to match current behavior, so we only change between the "-it" and "-i" option.
> 
> @snnn @edgchen1 @skottmckay FYI.

thanks for trying it out.

FYI, if that is the only problem, it's also possible to accept the licenses non-interactively: https://stackoverflow.com/a/45782695
`yes | sdkmanager --licenses`

File: tools/android_custom_build/build_custom_android_package.py
Comment (edgchen1) [inline tools/android_custom_build/build_custom_android_package.py:31]: general comment: nit: "psuedo" -> "pseudo"

File: tools/android_custom_build/build_custom_android_package.py
Comment (edgchen1) [inline tools/android_custom_build/build_custom_android_package.py:32]: consider something like `--disable_docker_pseudo_tty` with `action="store_false"`
https://docs.python.org/3/library/argparse.html#action

**Comment by yeziyqf1** [comment]
> 

Got it - would you prefer this approach over the `docker_pseudo_tty `approach? I'm open to both if the licensing issue is resolved. Just want to confirm your preferences😀

**Comment by edgchen1** [comment]
> > 
> 
> Got it - would you prefer this approach over the `docker_pseudo_tty `approach? I'm open to both if the licensing issue is resolved. Just want to confirm your preferences😀

I have a slight preference for this, if it works. the fewer command line options the better :)

**Comment by yeziyqf1** [comment]
> 
@edgchen1 I took a look at the solution, while it is not too clear with me where are we specifying this` yes | sdkmanager --licenses`, as the custom android build script is using [build_aar_package.py](https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/android/build_aar_package.py) script. I don't find a clear path yet, will reach out for a deeper discussion offline.

**Comment by yeziyqf1** [comment]
Discussed with @edgchen1 offline, the looping issue we saw previously in non-interactive environment are a result of intermittent skdmanager install times out. On later run, this issue does not occur. So, we agreed that removing the `-it` option when running docker is the preferred approach, as it does not introduce one more flag into the script. Now the change is ready for review.

File: tools/android_custom_build/build_custom_android_package.py
Comment (yeziyqf1) [reply tools/android_custom_build/build_custom_android_package.py]: I've addressed this in the commit [92b0197](https://github.com/microsoft/onnxruntime/pull/13704/commits/92b019772a0c1b90c614eb3124df8a8290db7cad), meanwhile, as we have decided to go with removing `-it` flag directly, this would not be part of the change, thus resolving this.

File: tools/android_custom_build/build_custom_android_package.py
Comment (yeziyqf1) [reply tools/android_custom_build/build_custom_android_package.py]: Addressed.

File: tools/android_custom_build/build_custom_android_package.py
Comment (yeziyqf1) [reply tools/android_custom_build/build_custom_android_package.py]: I've addressed this in the commit [92b0197](https://github.com/microsoft/onnxruntime/pull/13704/commits/92b019772a0c1b90c614eb3124df8a8290db7cad), meanwhile, as we have decided to go with removing -it flag directly, this would not be part of the change, thus resolving this.

**Review by edgchen1 — APPROVED** [review]

**Review by skottmckay — APPROVED** [review]

**Comment by edgchen1** [comment]
@snnn are the CI builds still expected to work on rel-1.12.1?

**Comment by ytaous** [comment]
@edgchen1 - should we continue with this PR?

**Comment by skottmckay** [comment]
Is there any need for it to be merged into an old release branch?

Also the script was recently updated and hopefully the new version does everything required with this line:

https://github.com/microsoft/onnxruntime/blob/e2dd1315c779f9f60aedd0243d61af625fdf5cc3/tools/android_custom_build/build_custom_android_package.py#L151

**Comment by ytaous** [comment]
> Is there any need for it to be merged into an old release branch?
> 
> Also the script was recently updated and hopefully the new version does everything required with this line:
> 
> https://github.com/microsoft/onnxruntime/blob/e2dd1315c779f9f60aedd0243d61af625fdf5cc3/tools/android_custom_build/build_custom_android_package.py#L151

I meant to continue with latest main branch.
@yeziyqf1 - I'll close it for now. pls feel free to re-open as needed.

**Comment by yeziyqf1** [comment]
> > Is there any need for it to be merged into an old release branch?
> > Also the script was recently updated and hopefully the new version does everything required with this line:
> > https://github.com/microsoft/onnxruntime/blob/e2dd1315c779f9f60aedd0243d61af625fdf5cc3/tools/android_custom_build/build_custom_android_package.py#L151
> 
> I meant to continue with latest main branch. @yeziyqf1 - I'll close it for now. pls feel free to re-open as needed.

@ytaous @skottmckay 
Hi all, we do need this change to merge in to the 1.12.1 branch, as this is the branch that we use in our release. The latest change in Android building script is awesome, which indeed solves the issue when we move to our next ONNXRuntime update, but this won't be reflected in the current Edge release, thus we still need to get this change merged into 1.12.1 branch. Thanks.

**Comment by yeziyqf1** [comment]
@edgchen1 @snnn If the CI verification pipeline now works fine again, we can now merge this PR.

**Comment by yeziyqf1** [comment]
@ytaous Could you help me reopen this PR? It seems like I don't have the permissions to do so.

**Comment by ytaous** [comment]
I don't think we have plan to back port a patch to 1.2.1, or create new release 1.2.2 for example.
you can't use later version?

**Comment by yeziyqf1** [comment]
> I don't think we have plan to back port a patch to 1.2.1, or create new release 1.2.2 for example. you can't use later version?

Edge's current supported ONNXRuntime version is 1.12.1, we use the 1.12.1 branch when creating the reduced operators build for ONNXRuntime Edge customers. We don't have plans to upgrade to a newer version currently. The decision to add this change to 1.12.1 has been discussed with Faith and Scott, and the PR is approved ready to merge, we are actively waiting for this PR to merge to unblock the Edge Android pipeline. Changming mentioned he was dedicated to fixing the CI issue with some cherry picks.

**Comment by yeziyqf1** [comment]
To unblock shipping Android ONNXRuntime 1.12.1 on Edge, we created downstream version of ONNXRuntime and make build pipeline utilize the downstream version instead. Thus, closing this PR.

**Pull request closed without merging**
